wallet/src/repository/wallet/mod.rs

```diff
@@ -648,116 +570,96 @@ where
-                Environment::Mainnet
-            });
-            let key_balance = model::KeyBalance { pkh, amount };
-
-            match db_utxo_set.insert(out_ptr.clone(), key_balance.clone()) {
-                None => {
-                    log::info!(
-                        "Found transaction to our address {}! Amount: +{} satowits",
-                        address,
-                        amount
-                    );
-                }
-                Some(x) => {
-                    if x != key_balance {
+        let mut output_amount: u64 = 0;
+        for (index, output) in txn.outputs.iter().enumerate() {
+            if let Some(model::Path { .. }) = self.db.get_opt(&keys::pkh(&output.pkh))? {
+                let address = output.pkh.bech32(if self.params.testnet {
+                    Environment::Testnet1
+                } else {
+                    Environment::Mainnet
+                });
+                let key_balance = model::KeyBalance {
+                    pkh: output.pkh,
+                    amount: output.value,
+                };
+
+                let out_ptr = model::OutPtr {
+                    txn_hash: Vec::from(txn_hash),
+                    output_index: index as u32,
+                };
+
+                match db_utxo_set.insert(out_ptr.clone(), key_balance.clone()) {
+                    Some(x) if x == key_balance => {}
+                    _ => {
                         log::info!(
                             "Found transaction to our address {}! Amount: +{} satowits",
                             address,
-                            amount
+                            output.value
                         );
                     }
                 }
+                state.utxo_set.insert(out_ptr, key_balance);
+
+                output_amount = output_amount
+                    .checked_add(output.value)
+                    .ok_or_else(|| Error::TransactionBalanceOverflow)?;
             }
+        }
 
-            batch.put(&keys::transaction_value(account, txn_id), amount)?;
-            batch.put(
-                keys::transaction_entry(account, txn_id),
-                model::TransactionEntry::Credit,
-            )?;
+        let txn_balance = (output_amount as i128)
+            .checked_sub(input_amount as i128)
+            .ok_or_else(|| Error::TransactionBalanceUnderflow)?;
+
+        let txn_balance =
+            i64::try_from(txn_balance).map_err(|_| Error::TransactionValueOverflow)?;
+
+        if txn_balance != 0 {
+            let new_balance = if txn_balance > 0 {
+                state
+                    .balance
+                    .checked_add(txn_balance.abs() as u64)
+                    .ok_or_else(|| Error::TransactionBalanceOverflow)?
+            } else {
+                state
+                    .balance
+                    .checked_sub(txn_balance.abs() as u64)
+                    .ok_or_else(|| Error::TransactionBalanceUnderflow)?
```

Comment:
I don't think this is a good error name. Is this the error that appears when trying to create a transaction with more value than the account balance? If so, it needs to be more descriptive. In the node that error is called `NoMoney`, and provides context about the account balance and the transaction value.

If this error should never happen (and only appears when the storage is corrupted or something), then I'm fine with renaming it as:

```suggestion
                    .ok_or_else(|| Error::TransactionBalanceNegative)?
```